python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -0,0 +1,716 @@
+        if os.path.exists(old_path):
+            # move the file across
+            try:
+                old_umask = os.umask(0)
+                try:
+                    bundle.ensure_folder_exists(os.path.dirname(path_to_cached_thumb))
+                    shutil.move(old_path, path_to_cached_thumb)
+                finally:
+                    os.umask(old_umask)
+            except:
+                # ignore any errors in the transfer
+                pass
+
+        return path_to_cached_thumb
+
+    @staticmethod
+    def _get_thumbnail_path_old(url, bundle):
+        """
+        March 2015 - Previous implementation of thumbnail caching logic.
+        This has now been replaced by a new, improved sharding algorithm.
+        In the interest of disk management, keep this method around so that
+        the new logic can attempt to over files over into the new scheme 
+        if at all possible.
+
+        :param url:     Path to a thumbnail
+        :param bundle:  App, Engine or Framework instance
+        :returns:       Path as a string.
+        """
+
+        url_obj = urlparse.urlparse(url)
+        url_path = url_obj.path
+        path_chunks = url_path.split("/")
+
+        CHUNK_LEN = 16
+
+        # post process the path
+        # old (pre-S3) style result:
+        # path_chunks: [ "", "thumbs", "1", "2", "2.jpg"]
+
+        # s3 result, form 1:
+        # path_chunks: [u'',
+        #               u'9902b5f5f336fae2fb248e8a8748fcd9aedd822e',
+        #               u'be4236b8f198ae84df2366920e7ee327cc0a567e',
+        #               u'render_0400_t.jpg']
+
+        # s3 result, form 2:
+        # path_chunks: [u'', u'thumbnail', u'api_image', u'150']
+
+        def _to_chunks(s):
+            #split the string 'abcdefghxx' into ['abcdefgh', 'xx']
+            chunks = []
+            for start in range(0, len(s), CHUNK_LEN):
+                chunks.append( s[start:start+CHUNK_LEN] )
+            return chunks
+
+        new_chunks = []
+        for folder in path_chunks[:-1]: # skip the file name
+            if folder == "":
+                continue
+            if len(folder) > CHUNK_LEN:
+                # long url path segment like 9902b5f5f336fae2fb248e8a8748fcd9aedd822e
+                # split it into chunks for 4
+                new_chunks.extend( _to_chunks(folder) )
+            else:
+                new_chunks.append(folder)
+
+        # establish the root path
+        cache_path_items = [bundle.cache_location, "thumbnails"]
+        # append the folders
+        cache_path_items.extend(new_chunks)
+        # and append the file name
+        # all sg thumbs are jpegs so append extension too - some url forms don't have this.
+        cache_path_items.append("%s.jpeg" % path_chunks[-1])
+
+        # join up the path
+        path_to_cached_thumb = os.path.join(*cache_path_items)
+
+        return path_to_cached_thumb
+
+    # default individual task priorities
```

Comment:
What's the priority order? I am guessing higher means "more important"?

To be consistent with the current implementation (and I think this behaviour also makes sense), I think priorities should  tweaked to this - does that make sense?

```
# default individual task priorities
# (higher value means more important)

# thumbnail checks are local disk checks and very 
# fast and are always carried out before any shotgun calls
_CHECK_THUMB_PRIORITY           = 50

# the sg schema is often useful to have as early on as possible
# sometimes other shotgun operations also need the shotgun schema
# (and it's typically also cached) so this call has a higher prio than the
# rest of the sg calls
_SG_DOWNLOAD_SCHEMA_PRIORITY    = 40

# next in priority order are all our standard shotgun calls
_EXECUTE_METHOD_PRIORITY        = 30
_SG_DELETE_PRIORITY             = 30
_SG_CREATE_PRIORITY             = 30
_SG_UPDATE_PRIORITY             = 30
_SG_FIND_PRIORITY               = 30

# thumbnails are downloaded last
_DOWNLOAD_THUMB_PRIORITY        = 20

```